modules/core/IPrim.h

```diff
@@ -99,10 +81,41 @@ namespace rt {
 		 * @return The name of the primitive
 		 */
 		DllExport std::string				getName(void) const { return m_name; }
-	
-	
+        /**
+		 * @brief Returns normal of the primitive. Flips the normal if it's been set to flip.
+         * @param ray Ray intersecting the primitive
+		 * @return The normalized normal of the primitive at the ray - primitive intersection point
+		 */
+        DllExport Vec3f				        getNormal(const Ray& ray) const { return m_flipped ? -doGetNormal(ray): doGetNormal(ray); }
+        /**
+        * @brief Returns the  normal vector of the primitive in the ray - primitive intersection point
+        * @note In contrast to the @ref doGetNormal() method, this methods takes into account the possible normal interpolation along the primitive
+        * @param ray Ray intersecting the primitive
+        * @retunrn The normalized normal of the primitive at the ray - primitive intersection point
+        */
+        DllExport Vec3f				        getShadingNormal(const Ray& ray) const { return m_flipped ? -doGetShadingNormal(ray): doGetShadingNormal(ray); }
+        /**
+		 * @brief Flips the normal of the primitive.
+		 */
+        DllExport void				flipNormal() { m_flipped = !m_flipped; }
+
+    protected:
+        /**
+        * @brief Returns the normal vector of the primitive in the ray - primitive intersection point
+        * @param ray Ray intersecting the primitive
+        * @return The normalized normal of the primitive at the ray - primitive intersection point
+        */
+        DllExport virtual Vec3f				doGetNormal(const Ray& ray) const = 0;
+        /**
+        * @brief Returns the  normal vector of the primitive in the ray - primitive intersection point
+        * @note In contrast to the @ref doGetNormal() method, this methods takes into account the possible normal interpolation along the primitive
+        * @param ray Ray intersecting the primitive
+        * @retunrn The normalized normal of the primitive at the ray - primitive intersection point
+        */
+        DllExport virtual Vec3f				doGetShadingNormal(const Ray& ray) const { return doGetNormal(ray); }
 	private:
-		const ptr_shader_t	m_pShader;			///< Pointer to the sahder, see @ref  IShader
-		std::string			m_name;				///< Optional name of the primitive
+		const ptr_shader_t	m_pShader;			///< Pointer to the shader, see @ref  IShader.
+		std::string			m_name;				///< Optional name of the primitive.
+		bool			    m_flipped;			///< Flag which helps decide whether to flip the normal or not.
```

Comment:
We should not forget to initialize it to _false_ by defaut
`bool m_flipped = false;`